### ISSUE TITLE
[Proposal] Introduce State Machine as a new sub-class of BlocBase

### ISSUE THREAD

**Issue by Pierre2tm** [felangel/bloc#3246]
Hello everyone! 👋

Thanks for the amazing job done with BLoC and I hope you will find this proposal interesting.

🚧Disclamer🚧

I'm not an expert in computer science, finite state automata, or bloc. What will come next is the result of my works on state machines. I present it here to serve as a base to be discussed.
It's also the first time I'm trying to contribute, I'm not used to this process so I hope I'm doing it the correct way.

# Context

This proposal aims to introduce a new class that inherits from `BlocBlase`, in addition to `Cubit` and `Bloc`: `StateMachine`.

Finite state machines are well known in computer science. There are different types of State machines covering different use cases. In this case, the term state machine will refer to a `Mealy Machine`. You can read more about [here](https://en.wikipedia.org/wiki/Mealy_machine).

A Mealy state machine is defined by a set of possible states and a set of possible transitions for each of them. The machine accepts input (UI event) and emits new states (or not) based on its current state and received input.

Where the state machine shines is in its expressiveness, predictability, and robustness:

- It makes it easy to know all possible states of the machine.
- It makes it easy to understand how state transit and what rules apply to these transitions.
- It guaranteed you to never be in an invalid state, which is especially valuable for sensitive parts of the app, like the payment funnel.

It's completely fitting with BLoC philosophy. It's helping you to know what state our application is in at any point in time and it's very predictable.

From app development perspective, State Machine comes with the following benefits:

* The state machine pattern eliminates bugs and weird situations because it won't let the UI transition to a state which we don’t know about.
* The machine is not accepting input which is not explicitly defined as acceptable for the current state. This eliminates the need for code that protects other code from execution.
* It forces the developer to think in a declarative way. This is because we have to define most of the logic upfront. BLoC already does that with events, the state machine pushes it further with states.

In addition to these qualities, State Machine is very close to what BLoC already does, which makes the integration of the state machine in BLoC package straightforward.

Both State Machine and BLoC received events from UI and emit new states based on their current states. The main differences reside in the design of their APIs.

- BLoC doesn't restrict you to a given set of States. You can emit an infinity of different states, as long as there inherit from the base State class. State machines force you to define every possible state at initialization.
- BLoC doesn't restrict you in the transition to a new state. At each event received, you can emit 0 or more new states. With a state machine, you can only emit **1** new State at a time.
- BLoC will react to an event at any time as long a handler has been defined with `on<>()`API. State Machine will **not** react to an event that is not defined for the **current** state.

|                    | BLoC                                | State Machine                                           |
| -------------------| ------------------------------------| --------------------------------------------------------|
| Set of States      | unrestricted                        | defined at initialization                               |
| Set of events      | defined at initialization           | defined at initialization                               |
| New State emission | 0 or more when an event is received | 0 or 1 when an event is received                        |
| Event reaction     | When an event is received           | When an event is received and defined for current state |

To sum up I would say BLoC is more focused on reacting to events while a state machine is focused on transiting inside a graph of states.

To go dive further into the subject, you can look at these great resources. These are javascript's react articles but the concept applies to dart as well. They explain in-depth why state machines are good for app state management.

Blog posts

- [You are managing state? Think twice.](https://krasimirtsonev.com/blog/article/managing-state-in-javascript-with-state-machines-stent)
- [The rise of state machine](https://www.smashingmagazine.com/2018/01/rise-state-machines/)
- [Robust React User Interfaces with Finite State Machines](https://css-tricks.com/robust-react-user-interfaces-with-finite-state-machines/)

Video

- [David Khourshid - Infinitely Better UIs with Finite Automata](https://www.youtube.com/watch?v=VU1NKX6Qkxc)

# Proposal 🚀

Considering the state of flutter state management ecosystem, the great similarity between BLoC and state machine, and the fact that there are already different types of blocs (Cubit and Bloc), I propose to add StateMachine as part of the bloc main package instead of creating a new state management package.

The implementation would be as close as possible to what bloc does to maintain consistency between different APIs.

StateMachine class will inherit from `BlocBase` class, that way it would beneficiated from the great bloc ecosystem (`flutter_bloc`, `bloc_test`, `hydrated_bloc`, `replay_bloc`, etc...).

```dart
abstract class StateMachine<Event, State> extends BlocBase<State>
```

StateMachine would use a builder pattern to easily define its underlying structure. [code_builder](https://pub.dev/packages/code_builder) from [fsm2](https://pub.dev/packages/fsm2) package are good sources of inspiration.

**note about fsm2**

fsm2 looks like a pretty good library for a state machine. It has advanced features like state nesting and visualization. I hadn't based my work on its code because it does not support state's data (an extended state in UML2 spec) but it has been a great source of inspiration.

##### StateMachine definition proposition

I came up with the following assertions:

- State machines can define N different states. All States should inherit from the same super type and should only be defined ones
- States can carry data.
- A state is defined by a list of possible transitions.
- A transition could or couldn't transit to a new state
- Transitions are evaluated sequentially, resulting in that the first transition to transit will emit the next state and subsequent transitions will not be evaluated.

The interface could look like that:

```dart
class TimerStateMachine extends StateMachine<TimerEvent, TimerState> {
  TimerStateMachine({required Ticker ticker})
      : _ticker = ticker,
        super(TimerInitial(_duration)) {
    define<TimerInitial>((b) => b
      ..on<TimerStarted>(
        (event, state) => TimerRunInProgress(event.duration),
      ));

    define<TimerRunInProgress>((b) => b
      ..onEnter((state) => _startTicker(state.duration))
      ..on<TimerTicked>(_onTicked)
      ..on<TimerPaused>(_onPaused)
      ..on<TimerReset>(_onReset));

    define<TimerRunPause>((b) => b
      ..on<TimerResumed>(_onResumed)
      ..on<TimerReset>(_onReset));

    define<TimerRunComplete>(
      (b) => b..on<TimerReset>(_onReset),
    );
  }
```

**Define**

Similar to Bloc's `on<Event>` API, `define<State>` let you define a state. It takes an optional builder at a parameter that gives the user opportunity to define the state's transition and side effects.

```dart
  void define<DefinedState extends State>([StateDefinitionBuilder Function(StateDefinitionBuilder)? definitionBuilder]);
```

`StateDefinitionBuilder` exposes `onEnter` and `onExit` that let users define a side effect that will be triggered when entering or exiting the state.

In addition, onEnter lets you directly change state, more on this in the transition section below.

**Transition**

A transition may or may not return a new state. If a state is returned, it's considered as state machine entering transition. If null is returned from transition, the next transition is evaluated. If all transitions returned null, state stay unchanged.

Transition can be async and will always be awaited before other transitions are evaluated. That way it let the user call an async function to decide if the transition will happen or not, from a repository by example.

A transition is defined using `on<Event>(builder)` API.

```dart
on<DefinedEvent>(
  FutureOr<State?> Function(DefinedEvent, DefinedState) builder,
);
```

In the example above, DefinedEvent is a sub-type of Event and DefinedState is a sub-type of State and it corresponds to the Type defined with `define<DefineState>`

In addition to the event transition, I propose to let `onEnter` be a transition. I called it *immediate transition*. This transition is always evaluated first when entering a state, and it's guaranteed that no event will be processed before the onEnter transition has finished being evaluated.

onEnter transition lets the user change state based on async code without the need to define an event. This is particularly useful for states that depend only on async code to change states.

Take this traffic lights state machine as an example:

```dart
class TrafficLights extends StateMachine<Event, State> {
  TrafficLight() : super(const Red()) {
    define<Green>((b) => b
      ..onEnter(
        (_) => Future.delayed(Duration(seconds: 25), () => Orange()),
      );

    define<Orange>((b) => b
      ..onEnter(
        (_) => Future.delayed(Duration(seconds: 5), () => Red()),
      );

    define<Red>((b) => b
      ..onEnter(
        (_) => Future.delayed(Duration(seconds: 30), () => Green()),
      ));
  }
}
```

It is also useful for initialization, where `Initialized` state is emitted after receiving data from an async func.

onEnter also lets you trigger side effects without changing state, you just need to return null to indicate that state will not change.

# Experimental Implementation

I've made an experimental package to test the feasibility of the concept. You can find it here: [bloc_state_machine_proposal](https://github.com/Pierre2tm/bloc_state_machine_proposal).

It features a working StateMachine that inherits from BlocBase and exposes API described above. you can import find the experimental package inside `/bloc_state_machine_experiment`.

I've also re-implemented the timer example with StateMachine, it can be found under `/example/flutter_timer_state_machine`

Of course, this isn't finished, but I think you could find this interesting and I hope it will contribute to the discussion thanks to a concrete example. Feel free to contribute to this experiment.

**Note about experiment implementation**

I've made two very opinionated choices with transition:

* made them async
* made onEnter a special case transition
  These choices impact underlying architecture and should be discussed, especially onEnter transition. I'm considering splitting onEnter in two, introducing an `immediateTransition` method in the builder, and keeping onEnter a side-effect only like onExit.

Transiting to a newState, where newState == currentState will trigger onEnter and onExit but will not trigger onChange and rebuild UI. This is because I don't override emit method and it's an opinionated design choice. It let self redirecting transition trigger onEnter/onExit side-effects, but I'm not sure having this behavior distinction with onChange is a good idea.

I've skipped transforming events and BlocOverrides in the experiment. I've noticed that BlocOverrides takes a bloc as parameter for onEvent, which makes StateMachine incompatible with it. However, this problem can be addressed by introducing a common superclass for Bloc and State Machine like

```dart
abstract class blocBaseEvent<Event, State> extends BlocBase<State> {}
```

I've also noticed that some tests of flutter_timer_state_machine example didn't pass, it looks like it's coming from `MockTicker`, I'm not able to explain why but I'm suspecting the missing lack of BlocOverride implementation be the cause.

# Additional Features

FSM2 has a nice feature called nested States. Take back our Traffic light example, what if we want to add an On and Off state? Green, Orange, and Red are On states, and adding Off transition in each of them is a lot of boilerplate.
Nested states solve this issue by enabling the following syntax:

```dart
class TrafficLights extends StateMachine<Event, State> {
  TrafficLight() : super(const Red()) {
    define<On>((b) => b
        ..on<TurnOff>((s, e) => Off())
        ..define<Green>((b) => b
          ..onEnter(
            (_) => Future.delayed(Duration(seconds: 25), () => Orange()),
          ));
        ..define<Orange>((b) => b
          ..onEnter(
            (_) => Future.delayed(Duration(seconds: 5), () => Red()),
          ));
        ..define<Red>((b) => b
          ..onEnter(
            (_) => Future.delayed(Duration(seconds: 30), () => Green()),
          ));
    );
    define<Off>((b) => b
      ..on<TurnOn>((e, s) => Red())
    )
  }
}
```

This feature isn't available in the experimental package I've made because I want to get your feedback before investing more time in this, but it's definitely something I want to have.

# Thanks for your attention

Please give this issue a 👍 if you support the proposal or a 👎 if you're against it. If you disagree with the proposal I would appreciate it if you could comment with your reasoning.

Thanks so much for giving attention to this proposal, and let's bring the power of state machines to flutter together! 🙏 🚀

**Comment by mtwichel** [comment]
Hey there 👋 I’ll share my 2 cents with the same caveat that I am no expert either and I love the time and thought that went into this proposal. It’s super encouraging too see this level of engagement with a package we all love 💙

After reading this proposal, my focus goes to the table of comparing state machines to existing BLoCs. You highlight four features:

- > Set of States | unrestricted | defined at initialization

Is the set of states **necessarily** unrestricted with BLoC? If you define an abstract state and have a finite state extend from it, doesn’t that restrict it to a finite set of states? For that matter, if you don’t need to carry data in your states, you can use an enum as your state and that is the finite set.

- > Set of events | defined at initialization | defined at initialization
    
These seem to be equivalent.

- > New State emission | 0 or more when an event is received | 0 or 1 when an event is received

While I see that BLoC doesn’t restrict to 0 or one states emitted when an event is received, even basic unit tests of your bloc should be able to identify any errors here. 

- > Event reaction | When an event is received | When an event is received and defined for current state

Again, you can write your blocs to only react to events given the state you’re in, and you can enforce this through unit testing, so I think this is possible with BLoC as well.

With all of these, it seems BLoC _can_ be equivalent to your proposal, but the `StateMachine` class provides a lot of syntactic sugar to make this pattern very easy to do and enforces some opinions about how a state machine should work. If I agree with those opinions (and honestly upon first reading they sound completely reasonable), what I like about BLoC is that it isn’t opinionated about **how** you structure your BLoCs. You get to embed your own opinions into how exactly they should work.

On the counter argument, you could say “this isn’t a breaking change and we’re just adding first class support to a very popular pattern. You can still do whatever you want with BLoC.” I think this point is a good one, but again my belief is that BLoC is a simple concept. I think adding this in to BLoC would needlessly confuse people (ie “is this what we’re supposed to use now? What is the advantages over existing stuff?”). It also introduces a new syntax that, while very similar to Blocs, it differs enough that I think it could confuse some people. BLoC is already (I think erroneously) criticized for being to confusing for new Flutter developers, and I am wary of anything that might contribute to that feeling.

Does that mean I think this idea should die? Not at all! I think this would be a super valuable package alongside the core bloc package. Heck, it could even be an officially supported package along with replay and hydrated bloc if the community loves it! It sounds like you have the features working already, so it sounds like it wouldn’t be much extra work? You could argue this would hurt discoverability (which it probably would), but if it was on the bloc website and mentioned in the main readme (and tweeted by Felix lol), I think it would get some traction. But, this way, people could choose to add it to their project if it made sense to them.

Just my thoughts though. Again I can convey how happy I am to see people think through these things and put so much effort into making the community better. Thank you 🙏 💙

**Comment by cmc5788** [comment]
I didn't see this directly addressed in the proposal: what would the pros and cons be of putting something like this in a separate package vs `bloc` core?

The reason I ask is that if `StateMachine` shows up one day alongside Bloc and Cubit, there ought to be a fairly straightforward way of explaining the use case for each subclass to a developer who now has three tools to reach for.

Or is that not the intention?

**Comment by Pierre2tm** [comment]
Thank you for all feedback and answers, I feel positive energy here!

> With all of these, it seems BLoC can be equivalent to your proposal,

You're right, **there is nothing you can do with StateMachine that can't be done using BLoC.** When I've researched about state machines I've surprised myself by asking me if BLoC hasn't been designed with state machine in mind so close they are. BLoC is more powerful and flexible, but with great power comes great responsibilities. A beginner I more likely to go wrong with BLoC rather than state machine because StateMachine is stable by default. By that, I mean that you can't be in an invalid state with state machines where with bloc you totally can.

I'll take my example. I came from getx and I moved to bloc because I was searching for a more structured solution. My controllers (where you find business logic in getx) became difficult to maintain over time. I loved BLoC and its way of declaring UI events. I agree with you, it's not so much boilerplate and it has structural benefits. But after using it daily I found myself having the same issue that I had with getx. When I use only one state class in my bloc code became difficult to maintain. I have a lot of `if (state is )` guard. It's exactly the problem described in the first part of this [article](https://krasimirtsonev.com/blog/article/managing-state-in-javascript-with-state-machines-stent). Was it a bad code from me? Sure. But could use a sm instead of a bloc have prevented this? Probably.

In the timer example, I've re-implemented its already has shown benefits. SM version greatly shows up inner logic and business rules inside the constructor. Its also removes the need for `if-state guards` and makes handlers very clear about how they will update the inner state because everything is typed.

> what would the pros and cons be of putting something like this in a separate package vs bloc core?

For the reasons mentioned above, I think `StateMachine` is more beginner-friendly and I would recommend new users of bloc package to use StateMachine instead of Bloc and keep Bloc for advanced usage. I think it will help them make more robust and maintainable apps and prevent them from going wrong. There is more chance that new users use StateMachine if it's part of the main package. But this is my personal opinion and I understand that bloc and sm are too close and can confuse people.

That said, I'm not against making it a separate package, especially if it's officially supported!

However, to make this work, I think it still would need minor changes in the main bloc package:

- Some variables are private inside of `BlocBase`, like `_emitted`and `_blocObserver` which make their usage outside of bloc library impossible. It's easily fixed with`@protected` annotation.
- `BlocObserver` takes `Bloc` as parameter inside `onEvent` and `onTransition` witch makes `StateMachine` incompatible. Again easily fixed by adding a super class for both sm and bloc.


Now the decision to put it or not inside the main package is up to you!

**Comment by cmc5788** [comment]
> For the reasons mentioned above, I think StateMachine is more beginner-friendly and I would recommend new users of bloc package to use StateMachine instead of Bloc and keep Bloc for advanced usage. I think it will help them make more robust and maintainable apps and prevent them from going wrong. There is more chance that new users use StateMachine if it's part of the main package. But this is my personal opinion and I understand that bloc and sm are too close and can confuse people.

Just my current feeling, but if the primary reason a developer might decide on when to use a state machine vs a bloc is based on "beginner" vs "advanced" usage, that sounds like an argument for separate packages.

Cubit and Bloc have distinct use cases that make clear sense to coexist within a single app. I can come up with some fairly easy rules of thumb about when to reach for a Cubit; for example, when complex async logic isn't required or is managed externally. An app may very well have good technical reasons to contain both cubits and blocs regardless of the personal preference of a developer based and based only on "best fit" criteria.

Are there some specific use cases you have in mind that could reasonably co-exist with Bloc and Cubit aside from developer experience level or preference? Problems that a state machine is the best tool to solve? Personally I think there might be some, just looking for your thoughts on this.

Playing devil's advocate, if there's not a clear use case for the coexistence it may be less confusing to developers to keep competing but equivalent solutions out of the same package.

**Comment by Pierre2tm** [comment]
> An app may very well have good technical reasons to contain both cubits and blocs regardless of the personal preference of a developer and based only on "best fit" criteria.

That's a really good point. I agree with you and also with what mtwichel has said. Thinking about it twice, it's true that sm is very opinionated in the way you should structure your logic and there are no clear use cases that differ enough from bloc (at least I can't find a good one).

Maybe after messing around with state machines, it will be more clear if there are some cases where it should be used in favor of bloc.
For the moment I will continue my work based on the experiment to make it a usable package and continue investigations. I've created https://github.com/Pierre2tm/state_machine_bloc repo for this purpose. Feel welcome to give feedback and help me if you like this project and want to contribute.

Thank you a lot for your high-quality feedback, that was interesting and challenging.

I will post updates here or maybe you want to close this issue already? I think we can keep it open a little bit more in case some people give additional feedback here.

**Comment by Tokenyet** [comment]
I like the design, and just want to mention there is a FSM discussion with current implementation. (https://github.com/felangel/bloc/issues/2526#issuecomment-928334495)

Although this is imo a good type of bloc, It's a subset of BLoC and lose more flexibility than Cubit. However, If this is a new package that combined BLoC and FSM, this is great for some applications and games 😆

**Comment by Pierre2tm** [comment]
> I like the design, and just want to mention there is a FSM discussion with current implementation. (https://github.com/felangel/bloc/issues/2526#issuecomment-928334495)

Thanks to point this out! That's a clever usage of freezed. That's close to what I'm trying to do. After reading that I'm asking myself if state_machine_bloc is worth doing. At least it's a very good example of how to implement this stuff.

> By the way, there is another [state machine package](https://pub.dev/packages/state_graph_bloc) using bloc, you might want to take a look 👀
It's look like I'm not the only one to had this idea. I will study this package and maybe reach the authors, thank you !

**Comment by Pierre2tm** [comment]
@Tokenyet Thank you so much for referring me to the comment about the state machine using freezed (and of course its author)!
I have totally rethought my approach based on this.

Now, `StateMachine` is just a little wrapper around `Bloc`. It still exposes the api described above, but it uses `on` under the hood. I think this is a much cleaner approach and this way it should be 100% compatible with bloc.

I just released version 0.0.1-dev.2 with nested state support 💪 and a clearer interface for side effects.

I will add unit tests and more examples before releasing 0.0.1 but I think it is already somewhat usable :).

**Comment by felangel** [comment]
Closing this for now since it looks like there are no actionable next steps in the context of this library 👍
